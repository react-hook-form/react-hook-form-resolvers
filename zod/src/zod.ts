import { appendErrors, transformToNestObject } from 'react-hook-form';
import * as z from 'zod';
import { convertArrayToPathName } from '@hookform/resolvers';
import type { Resolver } from './types';

const parseErrorSchema = (
  zodError: z.ZodError,
  validateAllFieldCriteria: boolean,
) => {
  if (zodError.isEmpty) {
    return {};
  }

  return zodError.errors.reduce<Record<string, any>>(
    (previous, { path, message, code: type }) => {
      const currentPath = convertArrayToPathName(path);

      return {
        ...previous,
        ...(path
          ? previous[currentPath] && validateAllFieldCriteria
            ? {
                [currentPath]: appendErrors(
                  currentPath,
                  validateAllFieldCriteria,
                  previous,
                  type,
                  message,
                ),
              }
            : {
                [currentPath]: previous[currentPath] || {
                  message,
                  type,
                  ...(validateAllFieldCriteria
                    ? {
                        types: { [type]: message || true },
                      }
                    : {}),
                },
              }
          : {}),
      };
    },
    {},
  );
};

export const zodResolver: Resolver = (
  schema,
  schemaOptions,
  { mode } = { mode: 'async' },
) => async (values, _, { criteriaMode }) => {
  try {
    const result =
      mode === 'async'
        ? await schema.parseAsync(values, schemaOptions)
        : schema.parse(values, schemaOptions);

    return { values: result, errors: {} };
  } catch (error) {
    return {
      values: {},
      errors: transformToNestObject(
        parseErrorSchema(error, criteriaMode === 'all'),
      ),
    };
  }
};
