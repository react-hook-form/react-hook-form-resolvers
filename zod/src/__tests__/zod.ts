import * as z from 'zod';
import { zodResolver } from '..';

const schema = z
  .object({
    username: z.string().regex(/^\w+$/).min(3).max(30),
    password: z.string().regex(/^[a-zA-Z0-9]{3,30}/),
    repeatPassword: z.string(),
    accessToken: z.union([z.string(), z.number()]).optional(),
    birthYear: z.number().min(1900).max(2013).optional(),
    email: z.string().email().optional(),
    tags: z.array(z.string()),
    enabled: z.boolean(),
  })
  .refine((data) => data.password === data.repeatPassword, {
    message: "Passwords don't match",
    path: ['confirm'], // set path of error
  });

describe('zodResolver', () => {
  it('should return values from zodResolver when validation pass', async () => {
    const data: z.infer<typeof schema> = {
      username: 'Doe',
      password: 'Password123',
      repeatPassword: 'Password123',
      birthYear: 2000,
      email: 'john@doe.com',
      tags: ['tag1', 'tag2'],
      enabled: true,
    };

    const parseAsyncSpy = jest.spyOn(schema, 'parseAsync');

    const result = await zodResolver(schema)(data, undefined, { fields: {} });

    expect(parseAsyncSpy).toHaveBeenCalledTimes(1);
    expect(result).toEqual({ errors: {}, values: data });
  });

  it('should return values from zodResolver with `mode: sync` when validation pass', async () => {
    const data: z.infer<typeof schema> = {
      username: 'Doe',
      password: 'Password123',
      repeatPassword: 'Password123',
      birthYear: 2000,
      email: 'john@doe.com',
      tags: ['tag1', 'tag2'],
      enabled: true,
    };

    const parseSpy = jest.spyOn(schema, 'parse');
    const parseAsyncSpy = jest.spyOn(schema, 'parseAsync');

    const result = await zodResolver(schema, undefined, { mode: 'sync' })(
      data,
      undefined,
      { fields: {} },
    );

    expect(parseSpy).toHaveBeenCalledTimes(1);
    expect(parseAsyncSpy).not.toHaveBeenCalled();
    expect(result).toEqual({ errors: {}, values: data });
  });

  it('should return a single error from zodResolver when validation fails', async () => {
    const data = {
      password: '___',
      email: '',
      birthYear: 'birthYear',
    };

    const result = await zodResolver(schema)(data, undefined, { fields: {} });

    expect(result).toMatchSnapshot();
  });

  it('should return a single error from zodResolver with `mode: sync` when validation fails', async () => {
    const data = {
      password: '___',
      email: '',
      birthYear: 'birthYear',
    };

    const parseSpy = jest.spyOn(schema, 'parse');
    const parseAsyncSpy = jest.spyOn(schema, 'parseAsync');

    const result = await zodResolver(schema, undefined, { mode: 'sync' })(
      data,
      undefined,
      { fields: {} },
    );

    expect(parseSpy).toHaveBeenCalledTimes(1);
    expect(parseAsyncSpy).not.toHaveBeenCalled();
    expect(result).toMatchSnapshot();
  });

  it('should return all the errors from zodResolver when validation fails with `validateAllFieldCriteria` set to true', async () => {
    const data = {
      password: '___',
      email: '',
      birthYear: 'birthYear',
    };

    const result = await zodResolver(schema)(data, undefined, {
      fields: {},
      criteriaMode: 'all',
    });

    expect(result).toMatchSnapshot();
  });

  it('should return all the errors from zodResolver when validation fails with `validateAllFieldCriteria` set to true and `mode: sync`', async () => {
    const data = {
      password: '___',
      email: '',
      birthYear: 'birthYear',
    };

    const result = await zodResolver(schema, undefined, { mode: 'sync' })(
      data,
      undefined,
      {
        fields: {},
        criteriaMode: 'all',
      },
    );

    expect(result).toMatchSnapshot();
  });
});
