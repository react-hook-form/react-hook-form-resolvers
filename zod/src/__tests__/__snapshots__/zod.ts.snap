// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`zodResolver should return a single error from zodResolver when validation fails 1`] = `
Object {
  "errors": Object {
    "birthYear": Object {
      "message": "Invalid input",
      "type": "invalid_union",
    },
    "confirm": Object {
      "message": "Passwords don't match",
      "type": "custom_error",
    },
    "email": Object {
      "message": "Invalid email",
      "type": "invalid_string",
    },
    "enabled": Object {
      "message": "Required",
      "type": "invalid_type",
    },
    "password": Object {
      "message": "Invalid",
      "type": "invalid_string",
    },
    "repeatPassword": Object {
      "message": "Required",
      "type": "invalid_type",
    },
    "tags": Object {
      "message": "Required",
      "type": "invalid_type",
    },
    "username": Object {
      "message": "Required",
      "type": "invalid_type",
    },
  },
  "values": Object {},
}
`;

exports[`zodResolver should return a single error from zodResolver with \`mode: sync\` when validation fails 1`] = `
Object {
  "errors": Object {
    "birthYear": Object {
      "message": "Invalid input",
      "type": "invalid_union",
    },
    "confirm": Object {
      "message": "Passwords don't match",
      "type": "custom_error",
    },
    "email": Object {
      "message": "Invalid email",
      "type": "invalid_string",
    },
    "enabled": Object {
      "message": "Required",
      "type": "invalid_type",
    },
    "password": Object {
      "message": "Invalid",
      "type": "invalid_string",
    },
    "repeatPassword": Object {
      "message": "Required",
      "type": "invalid_type",
    },
    "tags": Object {
      "message": "Required",
      "type": "invalid_type",
    },
    "username": Object {
      "message": "Required",
      "type": "invalid_type",
    },
  },
  "values": Object {},
}
`;

exports[`zodResolver should return all the errors from zodResolver when validation fails with \`validateAllFieldCriteria\` set to true 1`] = `
Object {
  "errors": Object {
    "birthYear": Object {
      "message": "Invalid input",
      "type": "invalid_union",
      "types": Object {
        "invalid_union": "Invalid input",
      },
    },
    "confirm": Object {
      "message": "Passwords don't match",
      "type": "custom_error",
      "types": Object {
        "custom_error": "Passwords don't match",
      },
    },
    "email": Object {
      "message": "Invalid email",
      "type": "invalid_string",
      "types": Object {
        "invalid_string": "Invalid email",
      },
    },
    "enabled": Object {
      "message": "Required",
      "type": "invalid_type",
      "types": Object {
        "invalid_type": "Required",
      },
    },
    "password": Object {
      "message": "Invalid",
      "type": "invalid_string",
      "types": Object {
        "invalid_string": "Invalid",
      },
    },
    "repeatPassword": Object {
      "message": "Required",
      "type": "invalid_type",
      "types": Object {
        "invalid_type": "Required",
      },
    },
    "tags": Object {
      "message": "Required",
      "type": "invalid_type",
      "types": Object {
        "invalid_type": "Required",
      },
    },
    "username": Object {
      "message": "Required",
      "type": "invalid_type",
      "types": Object {
        "invalid_type": "Required",
      },
    },
  },
  "values": Object {},
}
`;

exports[`zodResolver should return all the errors from zodResolver when validation fails with \`validateAllFieldCriteria\` set to true and \`mode: sync\` 1`] = `
Object {
  "errors": Object {
    "birthYear": Object {
      "message": "Invalid input",
      "type": "invalid_union",
      "types": Object {
        "invalid_union": "Invalid input",
      },
    },
    "confirm": Object {
      "message": "Passwords don't match",
      "type": "custom_error",
      "types": Object {
        "custom_error": "Passwords don't match",
      },
    },
    "email": Object {
      "message": "Invalid email",
      "type": "invalid_string",
      "types": Object {
        "invalid_string": "Invalid email",
      },
    },
    "enabled": Object {
      "message": "Required",
      "type": "invalid_type",
      "types": Object {
        "invalid_type": "Required",
      },
    },
    "password": Object {
      "message": "Invalid",
      "type": "invalid_string",
      "types": Object {
        "invalid_string": "Invalid",
      },
    },
    "repeatPassword": Object {
      "message": "Required",
      "type": "invalid_type",
      "types": Object {
        "invalid_type": "Required",
      },
    },
    "tags": Object {
      "message": "Required",
      "type": "invalid_type",
      "types": Object {
        "invalid_type": "Required",
      },
    },
    "username": Object {
      "message": "Required",
      "type": "invalid_type",
      "types": Object {
        "invalid_type": "Required",
      },
    },
  },
  "values": Object {},
}
`;
